src/3D/L3DSubMesh.cpp

```diff
@@ -161,23 +161,27 @@ void L3DSubMesh::Submit(uint8_t viewId, ShaderProgram& program, uint64_t state,
 	_vertexBuffer->Bind();
 
 	auto const& skins = _l3dMesh.GetSkins();
+
+	const Texture2D* lastTexture = nullptr;
 	for (auto it = _primitives.begin(); it != _primitives.end(); ++it)
 	{
 		const Primitive& prim = *it;
 
+		const Texture2D* texture = nullptr;
 		if (prim.skinID != 0xFFFFFFFF)
 		{
-			const Texture2D* texture = nullptr;
 			if (skins.find(prim.skinID) != skins.end())
 				texture = skins.at(prim.skinID).get();
 			else
 				texture = &Game::instance()->GetMeshPack().GetTexture(prim.skinID);
 
-			program.SetTextureSampler("s_diffuse", 0, *texture);
+			if (lastTexture != texture)
+				program.SetTextureSampler("s_diffuse", 0, *texture);
 		}
 
 		_indexBuffer->Bind(prim.indicesCount, prim.indicesOffset);
-		bgfx::submit(viewId, program.GetRawHandle(), 0, preserveState || std::next(it) != _primitives.end());
+		bgfx::submit(viewId, program.GetRawHandle(), 0, lastTexture == texture && (preserveState || std::next(it) != _primitives.end()));
```

Comment:
this line concerns me, the point of state preservation is to maintain the transforms and other state information, does this not screw with that?

Reply:
Texture changes qualify as state changes in bgfx it seems